Add Clang support to Kbuild

Do not upstream this patch. Clang behavior is likely to change
---
 Kbuild.include |   36 ++++++++++++++++++++++++++++++++++++
 Makefile.build |    1 +
 2 files changed, 37 insertions(+)

Index: linux/scripts/Kbuild.include
===================================================================
--- linux.orig/scripts/Kbuild.include	2012-09-06 14:07:04.890904871 -0400
+++ linux/scripts/Kbuild.include	2012-09-06 14:09:05.741385626 -0400
@@ -106,9 +106,44 @@
 as-instr = $(call try-run,\
 	printf "%b\n" "$(1)" | $(CC) $(KBUILD_AFLAGS) -c -xassembler -o "$$TMP" -,$(2),$(3))
 
+# Test for clang
+ifeq ($(shell $(CC) -v 2>/dev/null | grep -c "clang version"), 1)
+
 # cc-option
 # Usage: cflags-y += $(call cc-option,-march=winchip-c6,-march=i586)
+cc-option = $(call try-run,\
+	$(CC) $(KBUILD_CPPFLAGS) $(KBUILD_CFLAGS) $(1) -c -xc /dev/null -o "$$TMP" 2>&1 | grep -q "argument unused during compilation: '$(1)'",$(2),$(1))
 
+# cc-option-yn
+# Usage: flag := $(call cc-option-yn,-march=winchip-c6)
+cc-option-yn = $(call try-run,\
+	$(CC) $(KBUILD_CPPFLAGS) $(KBUILD_CFLAGS) $(1) -c -xc /dev/null -o "$$TMP" 2>&1 | grep -q "argument unused during compilation: '$(1)'",n,y)
+
+# cc-option-align
+# Prefix align with either -falign or -malign
+cc-option-align = $(subst -functions=0,,\
+	$(call cc-option,-falign-functions=0,-malign-functions=0))
+
+# cc-disable-warning
+# Usage: cflags-y += $(call cc-disable-warning,unused-but-set-variable)
+cc-disable-warning = $(call try-run,\
+	$(CC) $(KBUILD_CPPFLAGS) $(KBUILD_CFLAGS) -W$(strip $(1)) -c -xc /dev/null -o "$$TMP" 2>&1 | grep -q "argument unused during compilation: '$(1)'",,-Wno-$(strip $(1)))
+
+# cc-version and cc-fullversion are not used in clang supported targets
+
+# cc-version
+# Usage gcc-ver := $(call cc-version)
+cc-version = 0
+
+# cc-fullversion
+# Usage gcc-ver := $(call cc-fullversion)
+cc-fullversion = 0
+
+# Otherwise assume GCC or something compatible
+else
+
+# cc-option
+# Usage: cflags-y += $(call cc-option,-march=winchip-c6,-march=i586)
 cc-option = $(call try-run,\
 	$(CC) $(KBUILD_CPPFLAGS) $(KBUILD_CFLAGS) $(1) -c -xc /dev/null -o "$$TMP",$(1),$(2))
 
@@ -135,6 +170,7 @@
 # Usage gcc-ver := $(call cc-fullversion)
 cc-fullversion = $(shell $(CONFIG_SHELL) \
 	$(srctree)/scripts/gcc-version.sh -p $(CC))
+endif
 
 # cc-ifversion
 # Usage:  EXTRA_CFLAGS += $(call cc-ifversion, -lt, 0402, -O1)
Index: linux/scripts/Makefile.build
===================================================================
--- linux.orig/scripts/Makefile.build	2012-09-06 14:07:04.902904721 -0400
+++ linux/scripts/Makefile.build	2012-09-06 14:07:48.686354288 -0400
@@ -69,6 +69,7 @@
 warning-1 += -Wold-style-definition
 warning-1 += $(call cc-option, -Wmissing-include-dirs)
 warning-1 += $(call cc-option, -Wunused-but-set-variable)
+warning-1 += $(call cc-option, -Wunused-value)
 warning-1 += $(call cc-disable-warning, missing-field-initializers)
 
 warning-2 := -Waggregate-return
